### ISSUE TITLE
Require API keys for clients that hit our APIs over a certain rate limit

### ISSUE THREAD

**Issue by mekarpeles** [internetarchive/openlibrary#9435]
I'm strongly considering following github's example and having API access keys be required for Open Library APIs above some threshold of usage

* **Greater Visibility:** see list of API callers by account / volume
* **Improve Contactability:** Don't rely on useragent / headers being added in; make it required.
* **Easier administration:** Instead of blocking an IP, simply (temporarily) block the account; make it easy for patron services / those beyond engineering team to intervene
* **Promote Responsible Use:** Reduce exposure of potentially abusive behavior through more explicit contract
* **Automated Detection:** Automatically reply to high use patterns with warnings and explanations

It doesn't have to be all-or-nothing. We can phase it in and also have a lower-access mode below some threshold.

## Which APIs?

search.json, covers, books.json are 3 candidates

### Stakeholders

**Note:** Before [making a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) _or_ updating an existing one, please [ensure your branch is up to date](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch).

**Comment by Spaarsh** [comment]
@cdrini I stumbled upon this issue when I searched issues based on Theme: Security.

This feature seems to be of great vitality. I went through the api.py and this feature warrants for the following tasks to be done:
Phase 1 - 
1. API generation algorithm - a function that generated the API based on some logic
2. Modified Login - if the login function gets a user and password data, then go via the normal route, else look for the API key and verify it.
3. Modifying all the existing methods to check for API keys.

Once these have been done, on top of these things, other features can be developed on top of it:
Phase 2 -
1. Request tracking
2. Rate limiting
3. User Metrics

The Phase 1 is simple up to step 3. Rigorous testing would be needed to check if the methods have been correctly updated.

**Comment by Spaarsh** [comment]
@mekarpeles @jimchamp I filled out the volunteers form to become a part of the slack channel to discuss this further, but haven't heard back yet. I even tried to join the community call for this week but maybe it wasn't conducted?

**Comment by jimchamp** [comment]
@Spaarsh I see that you have been added to our Slack.  Maybe there's an invitation in your inbox?
Our community calls run from 0900-1000 PST every Tuesday (more information [here](https://github.com/internetarchive/openlibrary/wiki/Community-Call)).

**Comment by Spaarsh** [comment]
@jimchamp Yes! I am a part of the Slack channel now. Thank you! I will look at the community call timings as well. Thanks again. Additionally, is there any immediate action to be taken for this issue?

**Comment by mekarpeles** [comment]
We may be able to switch to a world where nginx reverse proxies to dockerized kong or tyk and then s3 `access_key` headers are used bucket / keep track of / rate limit requests...

But this requires us to have a fast way to lookup and validate someone's via our xauthn API and we can't currently do this in a performant way and doing so would require a large effort so I'm considering this issue blocked.
